src/freenet/client/filter/M3UFilter.java

```diff
@@ -0,0 +1,190 @@
+         * content type sniffing in the media tag and audio-players,
+         * so people could insert as text/plain to circumvent the
+         * filter.*/
+        String uriForcetypeMp3Suffix = "?type=audio/mpeg";
+        try (DataInputStream dis = new DataInputStream(input)) {
+            readcount = dis.read(nextbyte);
+            // read each line manually
+            while (readcount != -1) {
+                if (Arrays.equals(nextbyte, CHAR_COMMENT_START)) {
+                    isComment = true;
+                } else {
+                    isComment = false;
+                }
+                // read one line as a fileUri
+                numberOfDotsInUri = 0;
+                isBadUri = false;
+                fileIndex = 0;
+                fileUri = new byte[MAX_URI_LENGTH];
+                while (readcount != -1) {
+                    if (!isComment && 
+                        // do not include carriage return in filenames
+                        !Arrays.equals(nextbyte, CHAR_CARRIAGE_RETURN) &&
+                        // enforce maximum path length to avoid OOM attacks
+                        fileIndex <= MAX_URI_LENGTH) {
+                        fileUri[fileIndex] = nextbyte[0];
+                        fileIndex += readcount;
+                    }
+                    readcount = dis.read(nextbyte);
+                    if (Arrays.equals(nextbyte, CHAR_NEWLINE)) {
+                        if (!isComment) {
+                            // remove too long paths
+                            if (fileIndex <= MAX_URI_LENGTH) {
+                                for (int i = 0; i < fileIndex; i++) {
+                                    byte b = fileUri[i];
+                                    if (!isAllowedInUri(b)) {
+                                        isBadUri = true;
+                                        break;
+                                    } 
+                                    if ((byte)'.' == b) {
+                                        numberOfDotsInUri += 1;
+                                    }
+                                }
+                                if (numberOfDotsInUri != 1) {
+                                    isBadUri = true;
+                                }
+                                // use only the first fileIndex bytes
+                                String uri;
+                                // TODO: cleanly filter the URI, i.e. with processURI from GenericReadFilterCallback.
+                                if (isBadUri) {
+                                    uri = badUriReplacement;
+                                } else {
+                                    uri = new String(fileUri, 0, fileIndex, "UTF-8");
+                                    uri += uriForcetypeMp3Suffix;
+                                }
+                                output.write(uri.getBytes("UTF-8"));
+                                output.write(nextbyte);
```

Comment:
I'm not sure we should write straight back to the output bucket here... Do we want to throw if we find things we don't like... or mangle the output/truncate the file?